In [57]:
import pandas as pd
import numpy as np 
import matplotlib.pyplot as plt 
import seaborn as sns


In [58]:
df=pd.read_csv(r'C:\Users\MuntasirulHoque\OneDrive - ADA Global\Desktop\ML CODE\Coffee\Coffee-Quality\df_arabica_clean.csv')

In [59]:
df.head()

Unnamed: 0  ID Country of Origin                 Farm Name  \
0           0   0          Colombia          Finca El Paraiso   
1           1   1            Taiwan  Royal Bean Geisha Estate   
2           2   2              Laos        OKLAO coffee farms   
3           3   3        Costa Rica                 La Cumbre   
4           4   4          Colombia           Finca Santuario   

                                Lot Number                           Mill  \
0                               CQU2022015               Finca El Paraiso   
1  The 2022 Pacific Rim Coffee Summit,T037       Royal Bean Geisha Estate   
2  The 2022 Pacific Rim Coffee Summit,LA01  oklao coffee processing plant   
3                               CQU2022017        La Montana Tarrazu MIll   
4                               CQU2023002                Finca Santuario   

  ICO Number                   Company   Altitude                Region  ...  \
0        NaN      Coffee Quality Union  1700-1930        Piendamo,Cauca  ...   
1        NaN  Taiwan Coffee Laboratory       1200                Chiayi  ...   
2        NaN  Taiwan Coffee Laboratory       1300  Laos Borofen Plateau  ...   
3        NaN      Coffee Quality Union       1900    Los Santos,Tarrazu  ...   
4        NaN      Coffee Quality Union  1850-2100         Popayan,Cauca  ...   

  Total Cup Points  Moisture Percentage Category One Defects Quakers  \
0            89.33                 11.8                    0       0   
1            87.58                 10.5                    0       0   
2            87.42                 10.4                    0       0   
3            87.17                 11.8                    0       0   
4            87.08                 11.6                    0       2   

          Color Category Two Defects            Expiration  \
0         green                    3  September 21st, 2023   
1    blue-green                    0   November 15th, 2023   
2     yellowish                    2   November 15th, 2023   
3         green                    0  September 21st, 2023   
4  yellow-green                    2       March 5th, 2024   

                 Certification Body  \
0             Japan Coffee Exchange   
1  Taiwan Coffee Laboratory 台灣咖啡研究室   
2  Taiwan Coffee Laboratory 台灣咖啡研究室   
3             Japan Coffee Exchange   
4             Japan Coffee Exchange   

                               Certification Address  \
0  〒413-0002 静岡県熱海市伊豆山１１７３−５８ 1173-58 Izusan, Ata...   
1  QAHWAH CO., LTD 4F, No. 225, Sec. 3, Beixin Rd...   
2  QAHWAH CO., LTD 4F, No. 225, Sec. 3, Beixin Rd...   
3  〒413-0002 静岡県熱海市伊豆山１１７３−５８ 1173-58 Izusan, Ata...   
4  〒413-0002 静岡県熱海市伊豆山１１７３−５８ 1173-58 Izusan, Ata...   

                     Certification Contact  
0  松澤　宏樹　Koju Matsuzawa - +81(0)9085642901  
1     Lin, Jen-An Neil 林仁安 - 886-289116612  
2     Lin, Jen-An Neil 林仁安 - 886-289116612  
3  松澤　宏樹　Koju Matsuzawa - +81(0)9085642901  
4  松澤　宏樹　Koju Matsuzawa - +81(0)9085642901  

[5 rows x 41 columns]

In [60]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 207 entries, 0 to 206
Data columns (total 41 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Unnamed: 0             207 non-null    int64  
 1   ID                     207 non-null    int64  
 2   Country of Origin      207 non-null    object 
 3   Farm Name              205 non-null    object 
 4   Lot Number             206 non-null    object 
 5   Mill                   204 non-null    object 
 6   ICO Number             75 non-null     object 
 7   Company                207 non-null    object 
 8   Altitude               206 non-null    object 
 9   Region                 205 non-null    object 
 10  Producer               206 non-null    object 
 11  Number of Bags         207 non-null    int64  
 12  Bag Weight             207 non-null    object 
 13  In-Country Partner     207 non-null    object 
 14  Harvest Year           207 non-null    object 
 15  Gradin

In [61]:
df.shape

(207, 41)

Data is very small only 207 rows 
but data has a large number of features

In [62]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 207 entries, 0 to 206
Data columns (total 41 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Unnamed: 0             207 non-null    int64  
 1   ID                     207 non-null    int64  
 2   Country of Origin      207 non-null    object 
 3   Farm Name              205 non-null    object 
 4   Lot Number             206 non-null    object 
 5   Mill                   204 non-null    object 
 6   ICO Number             75 non-null     object 
 7   Company                207 non-null    object 
 8   Altitude               206 non-null    object 
 9   Region                 205 non-null    object 
 10  Producer               206 non-null    object 
 11  Number of Bags         207 non-null    int64  
 12  Bag Weight             207 non-null    object 
 13  In-Country Partner     207 non-null    object 
 14  Harvest Year           207 non-null    object 
 15  Gradin

In [63]:
df.describe()

Unnamed: 0          ID  Number of Bags       Aroma      Flavor  \
count  207.000000  207.000000      207.000000  207.000000  207.000000   
mean   103.000000  103.000000      155.449275    7.721063    7.744734   
std     59.899917   59.899917      244.484868    0.287626    0.279613   
min      0.000000    0.000000        1.000000    6.500000    6.750000   
25%     51.500000   51.500000        1.000000    7.580000    7.580000   
50%    103.000000  103.000000       14.000000    7.670000    7.750000   
75%    154.500000  154.500000      275.000000    7.920000    7.920000   
max    206.000000  206.000000     2240.000000    8.580000    8.500000   

       Aftertaste    Acidity        Body     Balance  Uniformity  Clean Cup  \
count  207.000000  207.00000  207.000000  207.000000  207.000000      207.0   
mean     7.599758    7.69029    7.640918    7.644058    9.990338       10.0   
std      0.275911    0.25951    0.233499    0.256299    0.103306        0.0   
min      6.670000    6.83000    6.830000    6.670000    8.670000       10.0   
25%      7.420000    7.50000    7.500000    7.500000   10.000000       10.0   
50%      7.580000    7.67000    7.670000    7.670000   10.000000       10.0   
75%      7.750000    7.87500    7.750000    7.790000   10.000000       10.0   
max      8.420000    8.58000    8.250000    8.420000   10.000000       10.0   

       Sweetness     Overall  Defects  Total Cup Points  Moisture Percentage  \
count      207.0  207.000000    207.0        207.000000           207.000000   
mean        10.0    7.676812      0.0         83.706570            10.735266   
std          0.0    0.306359      0.0          1.730417             1.247468   
min         10.0    6.670000      0.0         78.000000             0.000000   
25%         10.0    7.500000      0.0         82.580000            10.100000   
50%         10.0    7.670000      0.0         83.750000            10.800000   
75%         10.0    7.920000      0.0         84.830000            11.500000   
max         10.0    8.580000      0.0         89.330000            13.500000   

       Category One Defects     Quakers  Category Two Defects  
count            207.000000  207.000000            207.000000  
mean               0.135266    0.690821              2.251208  
std                0.592070    1.686918              2.950183  
min                0.000000    0.000000              0.000000  
25%                0.000000    0.000000              0.000000  
50%                0.000000    0.000000              1.000000  
75%                0.000000    1.000000              3.000000  
max                5.000000   12.000000             16.000000

In [64]:
df.drop(labels=["Unnamed: 0","ICO Number","ID"],axis=1,inplace=True)

In [65]:
from sklearn.impute import SimpleImputer

In [66]:
cat_imputer = SimpleImputer(strategy="most_frequent")

In [67]:
cat_list = df.select_dtypes(include=["object"]).columns.tolist()

In [68]:
df[cat_list] = cat_imputer.fit_transform(df[cat_list])

In [69]:
df.head()

Country of Origin                 Farm Name  \
0          Colombia          Finca El Paraiso   
1            Taiwan  Royal Bean Geisha Estate   
2              Laos        OKLAO coffee farms   
3        Costa Rica                 La Cumbre   
4          Colombia           Finca Santuario   

                                Lot Number                           Mill  \
0                               CQU2022015               Finca El Paraiso   
1  The 2022 Pacific Rim Coffee Summit,T037       Royal Bean Geisha Estate   
2  The 2022 Pacific Rim Coffee Summit,LA01  oklao coffee processing plant   
3                               CQU2022017        La Montana Tarrazu MIll   
4                               CQU2023002                Finca Santuario   

                    Company   Altitude                Region  \
0      Coffee Quality Union  1700-1930        Piendamo,Cauca   
1  Taiwan Coffee Laboratory       1200                Chiayi   
2  Taiwan Coffee Laboratory       1300  Laos Borofen Plateau   
3      Coffee Quality Union       1900    Los Santos,Tarrazu   
4      Coffee Quality Union  1850-2100         Popayan,Cauca   

                Producer  Number of Bags Bag Weight  ... Total Cup Points  \
0  Diego Samuel Bermudez               1      35 kg  ...            89.33   
1                    曾福森               1      80 kg  ...            87.58   
2             WU TAO CHI              19      25 kg  ...            87.42   
3    Santa Maria de Dota               1      22 kg  ...            87.17   
4       Camilo Merizalde               2      24 kg  ...            87.08   

  Moisture Percentage Category One Defects Quakers         Color  \
0                11.8                    0       0         green   
1                10.5                    0       0    blue-green   
2                10.4                    0       0     yellowish   
3                11.8                    0       0         green   
4                11.6                    0       2  yellow-green   

  Category Two Defects            Expiration  \
0                    3  September 21st, 2023   
1                    0   November 15th, 2023   
2                    2   November 15th, 2023   
3                    0  September 21st, 2023   
4                    2       March 5th, 2024   

                 Certification Body  \
0             Japan Coffee Exchange   
1  Taiwan Coffee Laboratory 台灣咖啡研究室   
2  Taiwan Coffee Laboratory 台灣咖啡研究室   
3             Japan Coffee Exchange   
4             Japan Coffee Exchange   

                               Certification Address  \
0  〒413-0002 静岡県熱海市伊豆山１１７３−５８ 1173-58 Izusan, Ata...   
1  QAHWAH CO., LTD 4F, No. 225, Sec. 3, Beixin Rd...   
2  QAHWAH CO., LTD 4F, No. 225, Sec. 3, Beixin Rd...   
3  〒413-0002 静岡県熱海市伊豆山１１７３−５８ 1173-58 Izusan, Ata...   
4  〒413-0002 静岡県熱海市伊豆山１１７３−５８ 1173-58 Izusan, Ata...   

                     Certification Contact  
0  松澤　宏樹　Koju Matsuzawa - +81(0)9085642901  
1     Lin, Jen-An Neil 林仁安 - 886-289116612  
2     Lin, Jen-An Neil 林仁安 - 886-289116612  
3  松澤　宏樹　Koju Matsuzawa - +81(0)9085642901  
4  松澤　宏樹　Koju Matsuzawa - +81(0)9085642901  

[5 rows x 38 columns]

In [70]:
df["Bag Weight"] = df["Bag Weight"].str.split().str.get(0).astype(int)

In [71]:
df.head()

Country of Origin                 Farm Name  \
0          Colombia          Finca El Paraiso   
1            Taiwan  Royal Bean Geisha Estate   
2              Laos        OKLAO coffee farms   
3        Costa Rica                 La Cumbre   
4          Colombia           Finca Santuario   

                                Lot Number                           Mill  \
0                               CQU2022015               Finca El Paraiso   
1  The 2022 Pacific Rim Coffee Summit,T037       Royal Bean Geisha Estate   
2  The 2022 Pacific Rim Coffee Summit,LA01  oklao coffee processing plant   
3                               CQU2022017        La Montana Tarrazu MIll   
4                               CQU2023002                Finca Santuario   

                    Company   Altitude                Region  \
0      Coffee Quality Union  1700-1930        Piendamo,Cauca   
1  Taiwan Coffee Laboratory       1200                Chiayi   
2  Taiwan Coffee Laboratory       1300  Laos Borofen Plateau   
3      Coffee Quality Union       1900    Los Santos,Tarrazu   
4      Coffee Quality Union  1850-2100         Popayan,Cauca   

                Producer  Number of Bags  Bag Weight  ... Total Cup Points  \
0  Diego Samuel Bermudez               1          35  ...            89.33   
1                    曾福森               1          80  ...            87.58   
2             WU TAO CHI              19          25  ...            87.42   
3    Santa Maria de Dota               1          22  ...            87.17   
4       Camilo Merizalde               2          24  ...            87.08   

  Moisture Percentage Category One Defects Quakers         Color  \
0                11.8                    0       0         green   
1                10.5                    0       0    blue-green   
2                10.4                    0       0     yellowish   
3                11.8                    0       0         green   
4                11.6                    0       2  yellow-green   

  Category Two Defects            Expiration  \
0                    3  September 21st, 2023   
1                    0   November 15th, 2023   
2                    2   November 15th, 2023   
3                    0  September 21st, 2023   
4                    2       March 5th, 2024   

                 Certification Body  \
0             Japan Coffee Exchange   
1  Taiwan Coffee Laboratory 台灣咖啡研究室   
2  Taiwan Coffee Laboratory 台灣咖啡研究室   
3             Japan Coffee Exchange   
4             Japan Coffee Exchange   

                               Certification Address  \
0  〒413-0002 静岡県熱海市伊豆山１１７３−５８ 1173-58 Izusan, Ata...   
1  QAHWAH CO., LTD 4F, No. 225, Sec. 3, Beixin Rd...   
2  QAHWAH CO., LTD 4F, No. 225, Sec. 3, Beixin Rd...   
3  〒413-0002 静岡県熱海市伊豆山１１７３−５８ 1173-58 Izusan, Ata...   
4  〒413-0002 静岡県熱海市伊豆山１１７３−５８ 1173-58 Izusan, Ata...   

                     Certification Contact  
0  松澤　宏樹　Koju Matsuzawa - +81(0)9085642901  
1     Lin, Jen-An Neil 林仁安 - 886-289116612  
2     Lin, Jen-An Neil 林仁安 - 886-289116612  
3  松澤　宏樹　Koju Matsuzawa - +81(0)9085642901  
4  松澤　宏樹　Koju Matsuzawa - +81(0)9085642901  

[5 rows x 38 columns]

In [72]:
df["Harvest Year"] = df["Harvest Year"].str.split("/").str.get(0).astype(int)

In [74]:
df2=df[["Harvest Year"]]

In [75]:
df2.head()

Harvest Year
0          2021
1          2021
2          2021
3          2022
4          2022

In [76]:
df2=df[["Grading Date"]]

In [79]:
df2.head()

Grading Date
0  September 21st, 2022
1   November 15th, 2022
2   November 15th, 2022
3  September 21st, 2022
4       March 6th, 2023

In [80]:
df["Grading Date"] = df["Grading Date"].str.split(",").str.get(1).astype(int)

In [81]:
df["Expiration"] = df["Expiration"].str.split(",").str.get(1).astype(int)

In [82]:
df["Color"].unique()

array(['green', 'blue-green', 'yellowish', 'yellow-green', 'yellow green',
       'greenish', 'brownish', 'yellow- green', 'browish-green',
       'bluish-green', 'pale yellow', 'yello-green'], dtype=object)

In [83]:
df.drop(labels=["Certification Contact", "Certification Address", "Lot Number", "Altitude", "Mill", "Status"], axis=1, 
        inplace=True)

In [84]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 207 entries, 0 to 206
Data columns (total 32 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Country of Origin     207 non-null    object 
 1   Farm Name             207 non-null    object 
 2   Company               207 non-null    object 
 3   Region                207 non-null    object 
 4   Producer              207 non-null    object 
 5   Number of Bags        207 non-null    int64  
 6   Bag Weight            207 non-null    int32  
 7   In-Country Partner    207 non-null    object 
 8   Harvest Year          207 non-null    int32  
 9   Grading Date          207 non-null    int32  
 10  Owner                 207 non-null    object 
 11  Variety               207 non-null    object 
 12  Processing Method     207 non-null    object 
 13  Aroma                 207 non-null    float64
 14  Flavor                207 non-null    float64
 15  Aftertaste            2

In [85]:
new_drop = ['Aroma', 'Flavor', 'Aftertaste', 'Acidity', 'Body', 'Balance', 
            'Uniformity', 'Clean Cup', 'Sweetness', 'Overall']

In [86]:
df.drop(labels=new_drop, axis=1, inplace=True)

In [88]:
corr_matrix = df.corr()
corr_matrix["Total Cup Points"].sort_values(ascending=False)

C:\Users\MuntasirulHoque\AppData\Local\Temp\ipykernel_13216\3814163101.py:1: FutureWarning: The default value of numeric_only in DataFrame.corr is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.
  corr_matrix = df.corr()


Total Cup Points        1.000000
Bag Weight              0.038214
Moisture Percentage    -0.046476
Category One Defects   -0.058156
Harvest Year           -0.127567
Grading Date           -0.193894
Expiration             -0.193894
Number of Bags         -0.243753
Category Two Defects   -0.314092
Quakers                -0.320307
Defects                      NaN
Name: Total Cup Points, dtype: float64

In [89]:
drop_attr = ["Defects", "Bag Weight", "Moisture Percentage", 
             "Category One Defects", "Harvest Year", 
             "Grading Date", "Expiration", "Number of Bags"]

df.drop(labels=drop_attr, axis=1, inplace=True)

In [90]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 207 entries, 0 to 206
Data columns (total 14 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Country of Origin     207 non-null    object 
 1   Farm Name             207 non-null    object 
 2   Company               207 non-null    object 
 3   Region                207 non-null    object 
 4   Producer              207 non-null    object 
 5   In-Country Partner    207 non-null    object 
 6   Owner                 207 non-null    object 
 7   Variety               207 non-null    object 
 8   Processing Method     207 non-null    object 
 9   Total Cup Points      207 non-null    float64
 10  Quakers               207 non-null    int64  
 11  Color                 207 non-null    object 
 12  Category Two Defects  207 non-null    int64  
 13  Certification Body    207 non-null    object 
dtypes: float64(1), int64(2), object(11)
memory usage: 22.8+ KB


In [91]:
df.isnull().sum()

Country of Origin       0
Farm Name               0
Company                 0
Region                  0
Producer                0
In-Country Partner      0
Owner                   0
Variety                 0
Processing Method       0
Total Cup Points        0
Quakers                 0
Color                   0
Category Two Defects    0
Certification Body      0
dtype: int64